pkg/proxy/nftables/proxier_test.go

```diff
@@ -507,27 +507,20 @@ func TestOverallNFTablesRules(t *testing.T) {
 		add chain ip kube-proxy filter-forward { type filter hook forward priority -101 ; }
 		add rule ip kube-proxy filter-forward ct state new jump endpoints-check
 		add rule ip kube-proxy filter-forward jump forward
-		add rule ip kube-proxy filter-forward ct state new jump firewall-check
 		add chain ip kube-proxy filter-input { type filter hook input priority -101 ; }
 		add rule ip kube-proxy filter-input ct state new jump endpoints-check
-		add rule ip kube-proxy filter-input ct state new jump firewall-check
+		add rule ip kube-proxy firewall-HVFWP5L3-ns5/svc5/tcp/p80 ip saddr { 203.0.113.0/25 } goto external-HVFWP5L3-ns5/svc5/tcp/p80 comment "LoadBalancer traffic allowed by LoadBalancerSourceRanges"
+		add rule ip kube-proxy firewall-HVFWP5L3-ns5/svc5/tcp/p80 drop comment "LoadBalancer traffic not accepted by LoadBalancerSourceRanges"
```

Comment:
Hm... in iptables you can't drop from the `nat` table... I would have said that nftables had the same restriction; you can't drop from a `type nat` chain... but maybe that's not true? Are you sure this actually works?

(FTR note that you can't actually test this from the e2e tests, because the `drop` never gets hit in normal operation on GCP. #85572)

Reply:
> Hm... in iptables you can't drop from the `nat` table... I would have said that nftables had the same restriction; you can't drop from a `type nat` chain... but maybe that's not true? Are you sure this actually works?

Yes, I have confirmed it works. A simple reproducer is as below:
```
nft -f - <<EOF
table ip test {
        chain test-chain {
                type nat hook output priority -101; policy accept;

                ip daddr 20.29.134.23 counter drop
        }
}
EOF
# It should timeout due to the drop action.
nc -vz 20.29.134.23 80 -w 3
# Check the counter, it drops 2 packets.
nft list ruleset
nft delete table ip test
```
Although there is a statement in [nftables wiki](https://wiki.nftables.org/wiki-nftables/index.php/Configuring_chains#Base_chain_types), it's mainly because subsequent packets bypass it, which is not a problem in this case as we handle the first packet only even in filter table:

> nat, which is used to perform Networking Address Translation (NAT). Only the first packet of a given flow hits this chain; subsequent packets bypass it. Therefore, never use this chain for filtering. The nat chain type is supported by the ip, ip6 and inet table families.

If you think it's not common to use nftables in this way, another way is to only jump to external chain (or remove firewall chain and add the rule directly to external chain as discussed above) when the source IP matches the ranges in nat table, and drops traffic matching loadBalancerIP via the existing set in filter table:
```
# nat
chain firewall-HVFWP5L3-ns5/svc5/tcp/p80 {
        ip saddr { 203.0.113.0/25 } goto external-HVFWP5L3-ns5/svc5/tcp/p80 comment "LoadBalancer traffic allowed by LoadBalancerSourceRanges"
}

# filter
chain firewall-check {
        ip daddr . meta l4proto . th dport @firewall drop
}
```

Reply:
Need to make sure it works for UDP too; for TCP if the first packet isn't allowed through then you can never establish a proper connection so it doesn't matter that we don't have a drop rule that would affect later packets. That may not be true for UDP.

> another way is

As I understand it, it _should_ be possible to make it work the way it's currently written (where the packets get dropped from a `type filter` chain before they get processed by the NAT chains) by using the right priorities. It's possible that I'm wrong about that though...

If there's no way to get the filter chain to run before the NAT chain, and your current approach turns out to not work for UDP, then we can still do it the way iptables currently does it, where we have the `fwChain` like in your current patch, but at the end it just doesn't do anything, and then a separate filter chain later on catches any traffic that still has the LB IP as its destination and drops it.

Reply:
I changed to follow the way it's currently written, linked it to prerouting chain to fix the issue, and changed the set to map to implement 2-stage match to avoid using concatenation of ranges, because no-endpoints filter still uses filter chain, just changing loadBalancerSourceRanges would look inconsistent.